[synse-server] Needs to create a clusterrolebinding for plugin descovery

On a fresh deploy of synse-server, logs will fail with tracebacks due to Forbidden errors from API

```
[2018-11-01 08:52:32 +0000] - (synse)[DEBUG] scan:31: Re-registering plugins
[2018-11-01 08:52:32 +0000] - (synse)[INFO] plugin:325: No plugin configurations for unix
[2018-11-01 08:52:32 +0000] - (synse)[DEBUG] plugin:327: unix plugin configuration: []
[2018-11-01 08:52:32 +0000] - (synse)[DEBUG] plugin:355: Registering plugins from default socket directory (/tmp/synse/procs)
[2018-11-01 08:52:32 +0000] - (synse)[INFO] plugin:380: Registered unix plugins: []
[2018-11-01 08:52:32 +0000] - (synse)[INFO] plugin:291: No plugin configurations for TCP
[2018-11-01 08:52:32 +0000] - (synse)[DEBUG] kubernetes:34: Using namespace "default" for k8s discovery
[2018-11-01 08:52:32 +0000] - (synse)[DEBUG] kubernetes:86: Using endpoint label selector: component=plugin
[2018-11-01 08:52:32 +0000] - (root)[ERROR] handlers:105: Traceback (most recent call last):
File "/usr/local/lib/python3.6/site-packages/sanic/app.py", line 603, in handle_request
response = await response
File "/usr/local/lib/python3.6/site-packages/synse/routes/core.py", line 55, in scan_route
response = await commands.scan(rack=rack, board=board, force=force)
File "/usr/local/lib/python3.6/site-packages/synse/commands/scan.py", line 32, in scan
plugin.register_plugins()
File "/usr/local/lib/python3.6/site-packages/synse/plugin.py", line 199, in register_plugins
addresses = kubernetes.discover()
File "/usr/local/lib/python3.6/site-packages/synse/discovery/kubernetes.py", line 43, in discover
addresses.extend(_register_from_endpoints(ns=ns, cfg=endpoints_cfg))
File "/usr/local/lib/python3.6/site-packages/synse/discovery/kubernetes.py", line 93, in _register_from_endpoints
endpoints = v1.list_namespaced_endpoints(namespace=ns, label_selector=label_selector)
File "/usr/local/lib/python3.6/site-packages/kubernetes/client/apis/core_v1_api.py", line 11782, in list_namespaced_endpoints
(data) = self.list_namespaced_endpoints_with_http_info(namespace, **kwargs)
File "/usr/local/lib/python3.6/site-packages/kubernetes/client/apis/core_v1_api.py", line 11885, in list_namespaced_endpoints_with_http_info
collection_formats=collection_formats)
File "/usr/local/lib/python3.6/site-packages/kubernetes/client/api_client.py", line 321, in call_api
_return_http_data_only, collection_formats, _preload_content, _request_timeout)
File "/usr/local/lib/python3.6/site-packages/kubernetes/client/api_client.py", line 155, in __call_api
_request_timeout=_request_timeout)
File "/usr/local/lib/python3.6/site-packages/kubernetes/client/api_client.py", line 342, in request
headers=headers)
File "/usr/local/lib/python3.6/site-packages/kubernetes/client/rest.py", line 231, in GET
query_params=query_params)
File "/usr/local/lib/python3.6/site-packages/kubernetes/client/rest.py", line 222, in request
raise ApiException(http_resp=r)
kubernetes.client.rest.ApiException: (403)
Reason: Forbidden
HTTP response headers: HTTPHeaderDict({'Content-Type': 'application/json', 'X-Content-Type-Options': 'nosniff', 'Date': 'Thu, 01 Nov 2018 08:52:32 GMT', 'Content-Length': '266'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"endpoints is forbidden: User \"system:serviceaccount:default:default\" cannot list endpoints in the namespace \"default\"","reason":"Forbidden","details":{"kind":"endpoints"},"code":403}
```

Since synse-server now uses Kubernetes API for plugin discovery, the chart will need to both create a service-account and a clusterrolebinding for read level permissions. Normally, a `Role` and `RoleBinding` would be sufficient, but since synse-server can technically be configured to search for endpoints not in their namespace, we'll use a `ClusterRole` and `ClusterRoleBinding`

I've not tested these templates, but it would look something like

## serviceaccount.yaml
```
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
{{ include "labels" . | indent 4 }}
  name: {{ template "fullname" . }}
```

## clusterrole.yaml

Since the URL being accessed is `/api/v1/namespaces/{namespace}/endpoints` we can decompose that into `apiGroups` "core", in one or more namespaces, for the endpoints resource. The method being invoked is `list` so we need at least the `list` and `get` verbs. I've added watch, because it's easy enough to include.

```
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: {{ template "fullname" . }}
rules:
- apiGroups: [""]                 # Empty string means the Core API group
  resources: ["endpoints"]
  verbs: ["get", "watch", "list"]
```

## clusterrolebinding.yaml
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: {{ template "fullname" . }}
  labels:
{{ include "labels" . | indent 4 }}
  annotations:
{{ include "annotations" . | indent 4 }}
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: {{ template "fullname" . }}
subjects:
- kind: ServiceAccount
  name: {{ template "fullname" . }}
  namespace: {{ .Release.Namespace }}
```

## deployment.yaml
Finally, the `Deployment` template spec will need to be updated to attach the ServiceAccount to synse-server.

Under `spec.template.spec` in the deployment add

```
serviceAccountName: {{ template "fullname" . }}
```

We should do our best to use `fullname` for these item to avoid collisions when multiple releases deployed in a single cluster / namespace.

--- a/synse-server/Chart.yaml
+++ b/synse-server/Chart.yaml
@@ -1,7 +1,7 @@
 ---
 
 name: synse-server
-version: 0.1.0
+version: 0.1.1
 appVersion: 2.2.4
 description: An HTTP API for the monitoring and control of physical and virtual devices.
 home: https://github.com/vapor-ware/synse-server

--- a/synse-server/templates/clusterrole.yaml
+++ b/synse-server/templates/clusterrole.yaml
@@ -1,0 +1,8 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: {{ template "fullname" . }}
+rules:
+- apiGroups: [""]                 # Empty string means Core API group
+  resources: ["endpoints"]        # Synse Server plugin discovery operates on Endpoint resources
+  verbs: ["get", "watch", "list"]

--- a/synse-server/templates/clusterrole.yaml
+++ b/synse-server/templates/clusterrole.yaml
@@ -1,3 +1,8 @@
+# If Synse Server is configured to run with plugin discovery via the
+# Kubernetes API, this template will be rendered. If the Kubernetes
+# discovery config is not set, Synse Server will not use the Kubernetes
+# API and will not need to be granted additional cluster permissions.
+{{- if .Values.config.plugin.discover.kubernetes }}
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
@@ -6,3 +11,4 @@ rules:
 - apiGroups: [""]                 # Empty string means Core API group
   resources: ["endpoints"]        # Synse Server plugin discovery operates on Endpoint resources
   verbs: ["get", "watch", "list"]
+{{- end }}

--- a/synse-server/templates/clusterrolebinding.yaml
+++ b/synse-server/templates/clusterrolebinding.yaml
@@ -1,3 +1,8 @@
+# If Synse Server is configured to run with plugin discovery via the
+# Kubernetes API, this template will be rendered. If the Kubernetes
+# discovery config is not set, Synse Server will not use the Kubernetes
+# API and will not need to be granted additional cluster permissions.
+{{- if .Values.config.plugin.discover.kubernetes }}
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRoleBinding
 metadata:
@@ -14,3 +19,4 @@ subjects:
 - kind: ServiceAccount
   name: {{ template "fullname" . }}
   namespace: {{ .Release.Namespace }}
+{{- end }}

--- a/synse-server/templates/clusterrolebinding.yaml
+++ b/synse-server/templates/clusterrolebinding.yaml
@@ -1,0 +1,16 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: {{ template "fullname" . }}
+  labels:
+{{ include "labels" . | indent 4 }}
+  annotations:
+{{ include "annotations" . | indent 4 }}
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: {{ template "fullname" . }}
+subjects:
+- kind: ServiceAccount
+  name: {{ template "fullname" . }}
+  namespace: {{ .Release.Namespace }}

--- a/synse-server/templates/deployment.yaml
+++ b/synse-server/templates/deployment.yaml
@@ -20,6 +20,7 @@ spec:
 {{ include "annotations" . | indent 8 }}
     spec:
       terminationGracePeriodSeconds: 3
+      serviceAccountName: {{ template "fullname" . }}
       volumes:
       - name: config
         configMap:

--- a/synse-server/templates/deployment.yaml
+++ b/synse-server/templates/deployment.yaml
@@ -20,7 +20,13 @@ spec:
 {{ include "annotations" . | indent 8 }}
     spec:
       terminationGracePeriodSeconds: 3
+      {{- if .Values.config.plugin.discover.kubernetes }}
+      # If Synse Server is configured to run with plugin discovery via the
+      # Kubernetes API, use the ServiceAccount that was created for it. The
+      # ServiceAccount will give Synse Server the proper cluster permissions
+      # to use the required Kubernetes API components.
       serviceAccountName: {{ template "fullname" . }}
+      {{- end }}
       volumes:
       - name: config
         configMap:

--- a/synse-server/templates/serviceaccount.yaml
+++ b/synse-server/templates/serviceaccount.yaml
@@ -1,0 +1,6 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ template "fullname" . }}
+  labels:
+{{ include "labels" . | indent 4 }}

--- a/synse-server/templates/serviceaccount.yaml
+++ b/synse-server/templates/serviceaccount.yaml
@@ -1,6 +1,12 @@
+# If Synse Server is configured to run with plugin discovery via the
+# Kubernetes API, this template will be rendered. If the Kubernetes
+# discovery config is not set, Synse Server will not use the Kubernetes
+# API and will not need to be granted additional cluster permissions.
+{{- if .Values.config.plugin.discover.kubernetes }}
 apiVersion: v1
 kind: ServiceAccount
 metadata:
   name: {{ template "fullname" . }}
   labels:
 {{ include "labels" . | indent 4 }}
+{{- end }}

